shared/bytesutil/bytes.go

```diff
@@ -353,5 +354,21 @@ func IsHex(b []byte) (bool, error) {
 	if b == nil {
 		return false, nil
 	}
-	return regexp.Match("^(0x)[0-9a-fA-F]+$", b)
+	r, err := regexp.Compile("^(0x)[0-9a-fA-F]+$")
+	if err != nil {
+		return false, errors.Wrapf(err, "could not parse regex")
+	}
+	return r.Match(b), nil
+}
+
+// IsHexOfLen checks whether the byte array is a hex number prefixed with '0x' and containing the required number of digits.
+func IsHexOfLen(b []byte, length uint64) (bool, error) {
+	if b == nil {
+		return false, nil
+	}
+	r, err := regexp.Compile("^(0x)[0-9a-fA-F]{" + strconv.FormatUint(length, 10) + "}$")
```

Comment:
You actually want to compile the regex outside of the function call, in a package-scope var. Compiling on each function call won't be meaningfully faster. The tradeoff there is that you can't perform the length check as part of the regex, since that isn't known ahead of time, which is why the modified regex in the benchmark only validates the character set (using `+` repeated and `$` end of string anchor) and separately checks the length of the byte slice.

Reply:
I see. This makes perfect sense.